File: src/code-samples/langchain/tool-return-direct.py
# :snippet-start: tool-return-direct-py
from langchain.agents import create_agent
from langchain.tools import tool
from langchain_openai import ChatOpenAI


@tool(return_direct=True)
def fetch_order_status(order_id: str) -> str:
    """Fetch the current status of a customer order."""
    # In production, query your order management system here
    return f"Order {order_id} is shipped and will arrive in 2 days."


agent = create_agent(
    ChatOpenAI(model="gpt-4o-mini"),
    tools=[fetch_order_status],
)

result = agent.invoke({
    "messages": [{"role": "user", "content": "What is the status of order #12345?"}]
})
# The agent returns the tool output directly without another LLM call:
# "Order 12345 is shipped and will arrive in 2 days."
# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    last = result["messages"][-1]
    text = (
        last.content_blocks[0]["text"]
        if getattr(last, "content_blocks", None)
        else str(last.content)
    )

    assert "shipped" in text and "12345" in text, text
    print("✓ returnDirect tool sample completed")
# :remove-end:

File: src/code-samples/langchain/tool-return-direct.ts
// :snippet-start: tool-return-direct-js
import { ChatOpenAI } from "@langchain/openai";
import { createAgent, tool } from "langchain";
import * as z from "zod";

const fetchOrderStatus = tool(
  ({ order_id }) => {
    return `Order ${order_id} is shipped and will arrive in 2 days.`;
  },
  {
    name: "fetch_order_status",
    description: "Fetch the current status of a customer order.",
    schema: z.object({ order_id: z.string() }),
    returnDirect: true,
  },
);

const agent = createAgent({
  model: new ChatOpenAI({ model: "gpt-4o-mini" }),
  tools: [fetchOrderStatus],
});

const result = await agent.invoke({
  messages: [
    { role: "user", content: "What is the status of order #12345?" },
  ],
});
// The agent returns the tool output directly without another LLM call:
// "Order 12345 is shipped and will arrive in 2 days."
// :snippet-end:

// :remove-start:
async function main() {
  if (!fetchOrderStatus.returnDirect) {
    throw new Error("Expected fetchOrderStatus.returnDirect to be true");
  }

  const last = result.messages.at(-1);
  const content =
    typeof last?.content === "string"
      ? last.content
      : JSON.stringify(last?.content);

  if (!content.includes("shipped") || !content.includes("12345")) {
    throw new Error(
      `Expected direct tool output for order 12345, got: ${content}`,
    );
  }

  console.log("✓ returnDirect tool sample completed");
}

main();
// :remove-end:
